src/main/java/org/jenkins/tools/test/PluginListerCli.java

```diff
@@ -0,0 +1,117 @@
+            description = "Comma-separated list of paths to external hooks JARs.",
+            converter = ExistingFileTypeConverter.class)
+    private Set<File> externalHooksJars = Set.of();
+
+    @CheckForNull
+    @CommandLine.Option(
+            names = {"-o", "--output"},
+            required = false,
+            description = "Location of the file to write containing the plugin and repositories.")
+    private File output;
+
+    @CheckForNull
+    @CommandLine.Option(
+            names = "--include-plugins",
+            split = ",",
+            arity = "1",
+            paramLabel = "plugin",
+            description =
+                    "Comma-separated list of plugin artifact IDs to test. If not set, every plugin in the WAR will be listed.")
+    private Set<String> includePlugins;
+
+    @CheckForNull
+    @CommandLine.Option(
+            names = "--exclude-plugins",
+            split = ",",
+            arity = "1",
+            paramLabel = "plugin",
+            description =
+                    "Comma-separated list of plugin artifact IDs to skip. If not set, only the plugins specified by --plugins will be listed (or all plugins otherwise).")
+    private Set<String> excludePlugins;
+
+    @Override
+    public Integer call() throws PluginCompatibilityTesterException {
+        List<PluginMetadataExtractor> metadataExtractors = PluginMetadataHooks.loadExtractors(externalHooksJars);
+
+        List<PluginMetadata> pluginMetadataList =
+                WarUtils.extractPluginMetadataFromWar(warFile, metadataExtractors, includePlugins, excludePlugins);
+
+        // Group the plugins into their actual repositories.
+        Map<String, List<PluginMetadata>> metaDataByRepoMap =
+                pluginMetadataList.stream().collect(Collectors.groupingBy(PluginMetadata::getGitUrl));
+
+        if (metaDataByRepoMap.isEmpty()) {
+            LOGGER.log(Level.WARNING, "Found no plugins in {0}", warFile);
+            return Integer.valueOf(5);
+        }
+
+        if (output != null) {
+            try (BufferedWriter writer = Files.newBufferedWriter(output.toPath())) {
+                for (Map.Entry<String, List<PluginMetadata>> entry : metaDataByRepoMap.entrySet()) {
+                    writer.write(formatEntry(entry));
+                    writer.write("\\n");
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        } else {
+            for (Map.Entry<String, List<PluginMetadata>> entry : metaDataByRepoMap.entrySet()) {
+                System.out.println(formatEntry(entry));
+            }
+        }
+        return Integer.valueOf(0);
+    }
+
+    private static String formatEntry(Entry<String, List<PluginMetadata>> entry) {
+        StringBuilder sb = new StringBuilder(entry.getKey());
+        for (PluginMetadata pm : entry.getValue()) {
+            sb.append("\t").append(pm.getPluginId());
+        }
+        return sb.toString();
+    }
```

Comment:
Interactive output of `list-plugins` is unreadable:

```
https://github.com/jenkinsci/workflow-multibranch-plugin.git    workflow-multibranch
https://github.com/jenkinsci/jakarta-activation-api-plugin.git  jakarta-activation-api
https://github.com/jenkinsci/text-finder-plugin.git     text-finder
https://github.com/jenkinsci/file-parameters-plugin.git file-parameters
https://github.com/jenkinsci/cloudbees-folder-plugin.git        cloudbees-folder
https://github.com/jenkinsci/github-branch-source-plugin.git    github-branch-source
https://github.com/jenkinsci/workflow-basic-steps-plugin.git    workflow-basic-steps
https://github.com/jenkinsci/ws-cleanup-plugin.git      ws-cleanup
https://github.com/jenkinsci/jersey2-api-plugin.git     jersey2-api
https://github.com/jenkinsci/instance-identity-plugin.git       instance-identity
https://github.com/jenkinsci/mina-sshd-api-plugin.git   mina-sshd-api-sftp      mina-sshd-api-scp       mina-sshd-api-common    mina-sshd-api-core
https://github.com/jenkinsci/javax-mail-api-plugin.git  javax-mail-api
https://github.com/jenkinsci/ionicons-api-plugin        ionicons-api
https://github.com/jenkinsci/script-security-plugin.git script-security
https://github.com/jenkinsci/copyartifact-plugin.git    copyartifact
https://github.com/jenkinsci/slack-plugin       slack
https://github.com/jenkinsci/ant-plugin.git     ant
https://github.com/jenkinsci/jackson2-api-plugin.git    jackson2-api
https://github.com/jenkinsci/aws-java-sdk-plugin.git    aws-java-sdk-elasticbeanstalk   aws-java-sdk-cloudformation     aws-java-sdk-sqs        aws-java-sdk-ssm        aws-java-sdk-codebuild  aws-java-sdk-logs       aws-java-sdk-efs      aws-java-sdk-iam        aws-java-sdk-minimal    aws-java-sdk-ecr        aws-java-sdk-ec2        aws-java-sdk-ecs        aws-java-sdk-sns
```

For it to be readable, it should have a heading and clearly defined column names. For example:

```
PLUGIN                    URL
jakarta-activation-api    https://github.com/jenkinsci/jakarta-activation-api-plugin.git
mina-sshd-api-common      https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-core        https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-scp         https://github.com/jenkinsci/mina-sshd-api-plugin.git
mina-sshd-api-sftp        https://github.com/jenkinsci/mina-sshd-api-plugin.git
workflow-multibranch      https://github.com/jenkinsci/workflow-multibranch-plugin.git
```

I do not care how the machine-parsable version looks, as long as it is clearly documented and consumers can rely on the machine-parsable format being stable.

Reply:
addressed in 8fad52370bf793321db77fe286792695efc960ea

output now uses spaces for consistent padding and follows the following example.

```
Apr 12, 2023 8:06:02 PM org.jenkins.tools.test.model.plugin_metadata.PluginMetadataHooks getPluginDetails
INFO: Extracting metadata about workflow-step-api
Apr 12, 2023 8:06:02 PM org.jenkins.tools.test.model.plugin_metadata.PluginMetadataHooks getPluginDetails
INFO: Extracting metadata about workflow-support
Plugin ID                                     Git Repository URL
ace-editor                                    https://github.com/jenkinsci/js-libs.git
active-directory                              https://github.com/jenkinsci/active-directory-plugin.git
analysis-model-api                            https://github.com/jenkinsci/analysis-model-api-plugin.git
ant                                           https://github.com/jenkinsci/ant-plugin.git
antisamy-markup-formatter                     https://github.com/jenkinsci/antisamy-markup-formatter-plugin.git
apache-httpcomponents-client-4-api            https://github.com/jenkinsci/apache-httpcomponents-client-4-api-plugin.git
artifact-manager-s3                           https://github.com/jenkinsci/artifact-manager-s3-plugin.git
authentication-tokens                         git://github.com/jenkinsci/authentication-tokens-plugin.git
aws-credentials                               git://github.com/jenkinsci/aws-credentials-plugin.git
aws-global-configuration                      git://github.com/jenkinsci/aws-global-configuration-plugin.git
aws-java-sdk-ec2                              https://github.com/jenkinsci/aws-java-sdk-plugin.git
aws-java-sdk-elasticbeanstalk                 https://github.com/jenkinsci/aws-java-sdk-plugin.git
aws-java-sdk-minimal                          https://github.com/jenkinsci/aws-java-sdk-plugin.git
blueocean-bitbucket-pipeline                  https://github.com/jenkinsci/blueocean-plugin.git
blueocean-commons                             https://github.com/jenkinsci/blueocean-plugin.git
blueocean-config                              https://github.com/jenkinsci/blueocean-plugin.git
blueocean-core-js                             https://github.com/jenkinsci/blueocean-plugin.git
blueocean-dashboard                           https://github.com/jenkinsci/blueocean-plugin.git
blueocean-display-url                         git://github.com/jenkinsci/blueocean-display-url-plugin.git
blueocean-events                              https://github.com/jenkinsci/blueocean-plugin.git
blueocean-git-pipeline                        https://github.com/jenkinsci/blueocean-plugin.git
```

Reply:
I cleaned this up even further.